File: app/fsharp/tests/Ch07/ResultHandlingTests.fs
module Ch07.ResultHandlingTests

open Xunit
open Ch07.ResultHandling
open Ch07.MusicArtists
open Ch07.PaymentMethods

// ============================================
// 基本的な Result のテスト
// ============================================

[<Fact>]
let ``tryParseInt は有効な整数文字列をパースする`` () =
    Assert.Equal(Ok 123, tryParseInt "123")
    Assert.Equal(Ok -456, tryParseInt "-456")

[<Fact>]
let ``tryParseInt は無効な文字列でエラーメッセージを返す`` () =
    match tryParseInt "abc" with
    | Error msg -> Assert.Contains("Can't parse", msg)
    | Ok _ -> Assert.Fail("Expected Error")

// ============================================
// TV番組パーサーのテスト（Result版）
// ============================================

[<Fact>]
let ``extractName は番組名を抽出する`` () =
    Assert.Equal(Ok "Breaking Bad", extractName "Breaking Bad (2008-2013)")

[<Fact>]
let ``extractName は無効な形式でエラーメッセージを返す`` () =
    match extractName "(2008-2013)" with
    | Error msg -> Assert.Contains("Can't extract name", msg)
    | Ok _ -> Assert.Fail("Expected Error")

[<Fact>]
let ``extractYearStart は開始年を抽出する`` () =
    Assert.Equal(Ok 2008, extractYearStart "Breaking Bad (2008-2013)")

[<Fact>]
let ``extractYearStart は無効な形式でエラーメッセージを返す`` () =
    match extractYearStart "Chernobyl (2019)" with
    | Error msg -> Assert.Contains("Can't extract start year", msg)
    | Ok _ -> Assert.Fail("Expected Error")

[<Fact>]
let ``parseShow は正常な形式をパースする`` () =
    match parseShow "Breaking Bad (2008-2013)" with
    | Ok show ->
        Assert.Equal("Breaking Bad", show.Title)
        Assert.Equal(2008, show.Start)
        Assert.Equal(2013, show.End)
    | Error _ -> Assert.Fail("Expected Ok")

[<Fact>]
let ``parseShow は単年の番組をパースする`` () =
    match parseShow "Chernobyl (2019)" with
    | Ok show ->
        Assert.Equal("Chernobyl", show.Title)
        Assert.Equal(2019, show.Start)
        Assert.Equal(2019, show.End)
    | Error _ -> Assert.Fail("Expected Ok")

[<Fact>]
let ``parseShow は無効な形式でエラーメッセージを返す`` () =
    match parseShow "(2008-2013)" with
    | Error msg -> Assert.Contains("Can't extract name", msg)
    | Ok _ -> Assert.Fail("Expected Error")

// ============================================
// バリデーションのテスト
// ============================================

[<Fact>]
let ``validateAge は有効な年齢を受け入れる`` () =
    Assert.Equal(Ok 25, validateAge 25)
    Assert.Equal(Ok 0, validateAge 0)
    Assert.Equal(Ok 150, validateAge 150)

[<Fact>]
let ``validateAge は負の年齢を拒否する`` () =
    match validateAge -5 with
    | Error msg -> Assert.Equal("Age cannot be negative", msg)
    | Ok _ -> Assert.Fail("Expected Error")

[<Fact>]
let ``validateAge は150歳より大きい年齢を拒否する`` () =
    match validateAge 200 with
    | Error msg -> Assert.Equal("Age cannot be greater than 150", msg)
    | Ok _ -> Assert.Fail("Expected Error")

[<Fact>]
let ``validateName は有効な名前を受け入れる`` () =
    Assert.Equal(Ok "Alice", validateName "Alice")

[<Fact>]
let ``validateName は空の名前を拒否する`` () =
    match validateName "" with
    | Error msg -> Assert.Equal("Name cannot be empty", msg)
    | Ok _ -> Assert.Fail("Expected Error")

[<Fact>]
let ``validateEmail は有効なメールアドレスを受け入れる`` () =
    Assert.Equal(Ok "test@example.com", validateEmail "test@example.com")

[<Fact>]
let ``validateEmail は @ がないメールアドレスを拒否する`` () =
    match validateEmail "invalid-email" with
    | Error msg -> Assert.Equal("Email must contain @", msg)
    | Ok _ -> Assert.Fail("Expected Error")

// ============================================
// MusicArtists のテスト
// ============================================

[<Fact>]
let ``wasArtistActive は StillActive のアーティストを正しく判定する`` () =
    let metallica = {
        Name = "Metallica"
        Genre = HeavyMetal
        Origin = US
        YearsActive = StillActive 1981
    }

    Assert.True(wasArtistActive metallica 1990 2000)
    Assert.True(wasArtistActive metallica 2000 2020)
    Assert.False(wasArtistActive metallica 1970 1980)

[<Fact>]
let ``wasArtistActive は ActiveBetween のアーティストを正しく判定する`` () =
    let ledZeppelin = {
        Name = "Led Zeppelin"
        Genre = HardRock
        Origin = England
        YearsActive = ActiveBetween (1968, 1980)
    }

    Assert.True(wasArtistActive ledZeppelin 1970 1975)
    Assert.True(wasArtistActive ledZeppelin 1975 1985)
    Assert.False(wasArtistActive ledZeppelin 1981 1990)

[<Fact>]
let ``activeLength は活動期間の長さを計算する`` () =
    let metallica = {
        Name = "Metallica"
        Genre = HeavyMetal
        Origin = US
        YearsActive = StillActive 1981
    }

    let ledZeppelin = {
        Name = "Led Zeppelin"
        Genre = HardRock
        Origin = England
        YearsActive = ActiveBetween (1968, 1980)
    }

    Assert.Equal(43, activeLength metallica 2024)
    Assert.Equal(12, activeLength ledZeppelin 2024)

[<Fact>]
let ``describeActivity は活動状態を説明する`` () =
    Assert.Equal("Active since 1981", describeActivity (StillActive 1981))
    Assert.Equal("Active from 1968 to 1980", describeActivity (ActiveBetween (1968, 1980)))

// ============================================
// 検索条件のテスト
// ============================================

[<Fact>]
let ``searchArtists はジャンルでフィルタする`` () =
    let artists = [
        { Name = "Metallica"; Genre = HeavyMetal; Origin = US; YearsActive = StillActive 1981 }
        { Name = "Beatles"; Genre = Pop; Origin = UK; YearsActive = ActiveBetween (1960, 1970) }
    ]

    let result = searchArtists artists [SearchByGenre [HeavyMetal]]

    Assert.Single(result) |> ignore
    Assert.Equal("Metallica", result.[0].Name)

[<Fact>]
let ``searchArtists は複数の条件でフィルタする`` () =
    let artists = [
        { Name = "Metallica"; Genre = HeavyMetal; Origin = US; YearsActive = StillActive 1981 }
        { Name = "Iron Maiden"; Genre = HeavyMetal; Origin = UK; YearsActive = StillActive 1975 }
        { Name = "Beatles"; Genre = Pop; Origin = UK; YearsActive = ActiveBetween (1960, 1970) }
    ]

    let conditions = [
        SearchByGenre [HeavyMetal]
        SearchByOrigin [UK]
    ]

    let result = searchArtists artists conditions

    Assert.Single(result) |> ignore
    Assert.Equal("Iron Maiden", result.[0].Name)

[<Fact>]
let ``searchArtists は活動期間でフィルタする`` () =
    let artists = [
        { Name = "Metallica"; Genre = HeavyMetal; Origin = US; YearsActive = StillActive 1981 }
        { Name = "Beatles"; Genre = Pop; Origin = UK; YearsActive = ActiveBetween (1960, 1970) }
    ]

    let result = searchArtists artists [SearchByActiveYears (1965, 1975)]

    // Beatles は 1960-1970 で活動、Metallica は 1981- なので Beatles のみ
    Assert.Single(result) |> ignore
    Assert.Equal("Beatles", result.[0].Name)

// ============================================
// PaymentMethods のテスト
// ============================================

[<Fact>]
let ``describePayment はクレジットカードを説明する`` () =
    let payment = CreditCard ("1234", "12/25")
    Assert.Equal("Credit card ending in 1234", describePayment payment)

[<Fact>]
let ``describePayment は銀行振込を説明する`` () =
    let payment = BankTransfer "9876"
    Assert.Equal("Bank transfer to account 9876", describePayment payment)

[<Fact>]
let ``describePayment は現金を説明する`` () =
    Assert.Equal("Cash payment", describePayment Cash)

[<Fact>]
let ``isValidPayment はクレジットカードを検証する`` () =
    Assert.True(isValidPayment (CreditCard ("1234", "12/25")))
    Assert.False(isValidPayment (CreditCard ("123", "12/25")))
    Assert.False(isValidPayment (CreditCard ("1234", "1225")))

[<Fact>]
let ``isValidPayment は銀行振込を検証する`` () =
    Assert.True(isValidPayment (BankTransfer "12345"))
    Assert.False(isValidPayment (BankTransfer ""))

[<Fact>]
let ``isValidPayment は現金を常に有効とする`` () =
    Assert.True(isValidPayment Cash)

File: app/csharp/tests/Ch07/ResultHandlingTests.cs
using Ch07;
using LanguageExt;
using static LanguageExt.Prelude;
using Xunit;
using static Ch07.ResultHandling;
using static Ch07.MusicArtists;
using static Ch07.PaymentMethods;

namespace GrokkingFP.CSharp.Tests.Ch07;

/// <summary>
/// 第7章: Either/Result 型のテスト
/// </summary>
public class ResultHandlingTests
{
    // ============================================
    // TryParseInt のテスト
    // ============================================

    [Fact]
    public void TryParseInt_ReturnsRightForValidInteger()
    {
        var result = ResultHandling.TryParseInt("42");
        Assert.True(result.IsRight);
        Assert.Equal(42, result.Match(Left: _ => 0, Right: r => r));
    }

    [Fact]
    public void TryParseInt_ReturnsLeftForInvalidInput()
    {
        var result = ResultHandling.TryParseInt("abc");
        Assert.True(result.IsLeft);
        Assert.Contains("Can't parse", result.Match(Left: l => l, Right: _ => ""));
    }

    // ============================================
    // TV番組パーサーのテスト（Either版）
    // ============================================

    [Fact]
    public void ExtractName_ReturnsRightForValidFormat()
    {
        var result = ExtractName("Breaking Bad (2008-2013)");
        Assert.True(result.IsRight);
        Assert.Equal("Breaking Bad", result.Match(Left: _ => "", Right: r => r));
    }

    [Fact]
    public void ExtractName_ReturnsLeftWithErrorMessage()
    {
        var result = ExtractName("No Brackets");
        Assert.True(result.IsLeft);
        Assert.Contains("Can't extract name", result.Match(Left: l => l, Right: _ => ""));
    }

    [Fact]
    public void ExtractYearStart_ReturnsRightForYearRange()
    {
        var result = ExtractYearStart("Breaking Bad (2008-2013)");
        Assert.True(result.IsRight);
        Assert.Equal(2008, result.Match(Left: _ => 0, Right: r => r));
    }

    [Fact]
    public void ExtractYearEnd_ReturnsRightForYearRange()
    {
        var result = ExtractYearEnd("Breaking Bad (2008-2013)");
        Assert.True(result.IsRight);
        Assert.Equal(2013, result.Match(Left: _ => 0, Right: r => r));
    }

    [Fact]
    public void ExtractSingleYear_ReturnsRightForSingleYear()
    {
        var result = ExtractSingleYear("Chernobyl (2019)");
        Assert.True(result.IsRight);
        Assert.Equal(2019, result.Match(Left: _ => 0, Right: r => r));
    }

    [Fact]
    public void ParseShow_ParsesYearRangeWithEither()
    {
        var result = ParseShow("Breaking Bad (2008-2013)");
        Assert.True(result.IsRight);
        var show = result.Match(Left: _ => (TvShow?)null, Right: s => s);
        Assert.NotNull(show);
        Assert.Equal("Breaking Bad", show!.Title);
        Assert.Equal(2008, show.Start);
        Assert.Equal(2013, show.End);
    }

    [Fact]
    public void ParseShow_ParsesSingleYearWithEither()
    {
        var result = ParseShow("Chernobyl (2019)");
        Assert.True(result.IsRight);
        var show = result.Match(Left: _ => (TvShow?)null, Right: s => s);
        Assert.NotNull(show);
        Assert.Equal("Chernobyl", show!.Title);
        Assert.Equal(2019, show.Start);
        Assert.Equal(2019, show.End);
    }

    [Fact]
    public void ParseShow_ReturnsLeftWithErrorForInvalidFormat()
    {
        var result = ParseShow("Invalid");
        Assert.True(result.IsLeft);
    }

    // ============================================
    // OrElse のテスト
    // ============================================

    [Fact]
    public void OrElse_ReturnsFirstWhenRight()
    {
        var first = Right<string, int>(42);
        var alternative = Right<string, int>(100);
        Assert.Equal(42, first.OrElse(alternative).Match(Left: _ => 0, Right: r => r));
    }

    [Fact]
    public void OrElse_ReturnsAlternativeWhenLeft()
    {
        var first = Left<string, int>("error");
        var alternative = Right<string, int>(100);
        Assert.Equal(100, first.OrElse(alternative).Match(Left: _ => 0, Right: r => r));
    }

    // ============================================
    // バリデーションのテスト
    // ============================================

    [Fact]
    public void ValidateAge_ReturnsRightForValidAge()
    {
        Assert.True(ValidateAge(25).IsRight);
        Assert.True(ValidateAge(0).IsRight);
        Assert.True(ValidateAge(150).IsRight);
    }

    [Fact]
    public void ValidateAge_ReturnsLeftForInvalidAge()
    {
        var negativeResult = ValidateAge(-1);
        Assert.True(negativeResult.IsLeft);
        Assert.Contains("negative", negativeResult.Match(Left: l => l, Right: _ => ""));

        var tooOldResult = ValidateAge(151);
        Assert.True(tooOldResult.IsLeft);
        Assert.Contains("150", tooOldResult.Match(Left: l => l, Right: _ => ""));
    }

    [Fact]
    public void ValidateName_ReturnsRightForValidName()
    {
        Assert.True(ValidateName("Alice").IsRight);
        Assert.True(ValidateName("Bob").IsRight);
    }

    [Fact]
    public void ValidateName_ReturnsLeftForInvalidName()
    {
        var emptyResult = ValidateName("");
        Assert.True(emptyResult.IsLeft);
        Assert.Contains("empty", emptyResult.Match(Left: l => l, Right: _ => ""));

        var tooLongResult = ValidateName(new string('a', 101));
        Assert.True(tooLongResult.IsLeft);
        Assert.Contains("100", tooLongResult.Match(Left: l => l, Right: _ => ""));
    }

    [Fact]
    public void ValidateEmail_ReturnsRightForValidEmail()
    {
        Assert.True(ValidateEmail("test@example.com").IsRight);
    }

    [Fact]
    public void ValidateEmail_ReturnsLeftForInvalidEmail()
    {
        var emptyResult = ValidateEmail("");
        Assert.True(emptyResult.IsLeft);

        var noAtResult = ValidateEmail("invalid.email");
        Assert.True(noAtResult.IsLeft);
        Assert.Contains("@", noAtResult.Match(Left: l => l, Right: _ => ""));
    }
}

/// <summary>
/// 音楽アーティストのテスト
/// </summary>
public class MusicArtistsTests
{
    private readonly Artist _metallica = new Artist(
        "Metallica",
        MusicGenre.HeavyMetal,
        new US(),
        new StillActive(1981)
    );

    private readonly Artist _beatles = new Artist(
        "The Beatles",
        MusicGenre.Pop,
        new UK(),
        new ActiveBetween(1960, 1970)
    );

    // ============================================
    // パターンマッチングのテスト
    // ============================================

    [Fact]
    public void WasArtistActive_ChecksStillActiveArtist()
    {
        Assert.True(WasArtistActive(_metallica, 2000, 2020));
        Assert.True(WasArtistActive(_metallica, 1980, 1985));
        Assert.False(WasArtistActive(_metallica, 1970, 1980));
    }

    [Fact]
    public void WasArtistActive_ChecksActiveBetweenArtist()
    {
        Assert.True(WasArtistActive(_beatles, 1965, 1968));
        Assert.True(WasArtistActive(_beatles, 1955, 1965));
        Assert.False(WasArtistActive(_beatles, 1971, 1980));
    }

    [Fact]
    public void ActiveLength_CalculatesForStillActive()
    {
        Assert.Equal(43, ActiveLength(_metallica, 2024));
    }

    [Fact]
    public void ActiveLength_CalculatesForActiveBetween()
    {
        Assert.Equal(10, ActiveLength(_beatles, 2024));
    }

    [Fact]
    public void DescribeActivity_DescribesStillActive()
    {
        var description = DescribeActivity(new StillActive(1981));
        Assert.Equal("Active since 1981", description);
    }

    [Fact]
    public void DescribeActivity_DescribesActiveBetween()
    {
        var description = DescribeActivity(new ActiveBetween(1960, 1970));
        Assert.Equal("Active from 1960 to 1970", description);
    }

    // ============================================
    // 検索のテスト
    // ============================================

    [Fact]
    public void MatchesCondition_MatchesByGenre()
    {
        var condition = new SearchByGenre(Seq(MusicGenre.HeavyMetal, MusicGenre.HardRock));
        Assert.True(MatchesCondition(_metallica, condition));
        Assert.False(MatchesCondition(_beatles, condition));
    }

    [Fact]
    public void MatchesCondition_MatchesByOrigin()
    {
        var condition = new SearchByOrigin(Seq<Location>(new US()));
        Assert.True(MatchesCondition(_metallica, condition));
        Assert.False(MatchesCondition(_beatles, condition));
    }

    [Fact]
    public void MatchesCondition_MatchesByActiveYears()
    {
        var condition = new SearchByActiveYears(1965, 1968);
        Assert.True(MatchesCondition(_beatles, condition));
        Assert.False(MatchesCondition(_metallica, condition));
    }

    [Fact]
    public void SearchArtists_FiltersWithMultipleConditions()
    {
        var artists = Seq(_metallica, _beatles);
        var conditions = Seq<SearchCondition>(
            new SearchByGenre(Seq(MusicGenre.HeavyMetal)),
            new SearchByOrigin(Seq<Location>(new US()))
        );
        var result = SearchArtists(artists, conditions);
        Assert.Equal(1, result.Count);
        Assert.Equal("Metallica", result.ToList().First().Name);
    }
}

/// <summary>
/// 支払い方法のテスト
/// </summary>
public class PaymentMethodsTests
{
    [Fact]
    public void DescribePayment_DescribesCreditCard()
    {
        var card = new CreditCard("1234567890123456", "12/25");
        var description = DescribePayment(card);
        Assert.Contains("Credit card", description);
        Assert.Contains("1234567890123456", description);
    }

    [Fact]
    public void DescribePayment_DescribesBankTransfer()
    {
        var transfer = new BankTransfer("123-456-789");
        var description = DescribePayment(transfer);
        Assert.Contains("Bank transfer", description);
        Assert.Contains("123-456-789", description);
    }

    [Fact]
    public void DescribePayment_DescribesCash()
    {
        var cash = new Cash();
        var description = DescribePayment(cash);
        Assert.Equal("Cash payment", description);
    }

    [Fact]
    public void IsValidPayment_ValidatesCreditCard()
    {
        Assert.True(IsValidPayment(new CreditCard("1234", "12/25")));
        Assert.False(IsValidPayment(new CreditCard("123", "12/25"))); // too short
        Assert.False(IsValidPayment(new CreditCard("1234", "1225"))); // no slash
    }

    [Fact]
    public void IsValidPayment_ValidatesBankTransfer()
    {
        Assert.True(IsValidPayment(new BankTransfer("123")));
        Assert.False(IsValidPayment(new BankTransfer("")));
    }

    [Fact]
    public void IsValidPayment_ValidatesCash()
    {
        Assert.True(IsValidPayment(new Cash()));
    }
}
